services/hanke-service/src/integrationTest/kotlin/fi/hel/haitaton/hanke/hakemus/HakemusServiceITest.kt

```diff
@@ -72,380 +72,749 @@ class HakemusServiceITest(
+                        .saveWithYhteystiedot { hakija() }
+                val hakemus = hakemusService.hakemusResponse(entity.id!!)
+                val yhteystieto = hakemusyhteystietoRepository.findAll().first()
+                val kayttaja =
+                    hankekayttajaRepository
+                        .findByHankeIdAndSahkopostiIn(hanke.id, listOf(KAYTTAJA_INPUT_HAKIJA.email))
+                        .single()
+                val newKayttaja = hankeKayttajaFactory.saveUser(hanke.id)
+                val originalAuditLogSize = auditLogRepository.findByType(ObjectType.HAKEMUS).size
+                val request =
+                    hakemus
+                        .toUpdateRequest()
+                        .withCustomerWithContactsRequest(
+                            CustomerType.COMPANY,
+                            yhteystieto.id,
+                            kayttaja.id,
+                            newKayttaja.id
+                        )
+                        .withWorkDescription("New work description")
+
+                val updatedHakemus = hakemusService.updateHakemus(hakemus.id, request, USERNAME)
+
+                assertThat(updatedHakemus.applicationData as JohtoselvitysHakemusDataResponse).all {
+                    prop(JohtoselvitysHakemusDataResponse::workDescription)
+                        .isEqualTo("New work description")
+                    prop(JohtoselvitysHakemusDataResponse::customerWithContacts)
+                        .isNotNull()
+                        .prop(CustomerWithContactsResponse::contacts)
+                        .transform { it.map { contact -> contact.hankekayttajaId } }
+                        .containsExactlyInAnyOrder(kayttaja.id, newKayttaja.id)
+                }
+                val applicationLogs = auditLogRepository.findByType(ObjectType.HAKEMUS)
+                assertThat(applicationLogs).hasSize(originalAuditLogSize + 1)
+            }
         }
     }
 
     @Nested
-    inner class UpdateHakemus {
-
-        private val intersectingArea =
-            ApplicationFactory.createApplicationArea(
-                name = "area",
-                geometry =
-                    "/fi/hel/haitaton/hanke/geometria/intersecting-polygon.json".asJsonResource()
-            )
-
-        private val notInHankeArea =
-            ApplicationFactory.createApplicationArea(
-                name = "area",
-                geometry = GeometriaFactory.polygon
-            )
-
-        @Test
-        fun `throws exception when the application does not exist`() {
-            assertThat(applicationRepository.findAll()).isEmpty()
-            val request =
-                HakemusUpdateRequestFactory.createFilledJohtoselvityshakemusUpdateRequest()
-
-            val exception = assertFailure { hakemusService.updateHakemus(1234, request, USERNAME) }
-
-            exception.hasClass(ApplicationNotFoundException::class)
-        }
+    inner class Kaivuilmoitushakemus {
```

Comment:
The application is called just `Kaivuilmoitus`. Legally, it's not an application, but more like a notification.

Reply:
Is `Johtoselvityshakemus` then just `Johtoselvitys`? It has sounded a bit redundant in my ears to have both substantives in the name. `Johtoselvitys` and `Kaivuilmoitus` sound more coherent.

Reply:
And does the same apply to `CableReportApplication`? Should there be just `Application` <-- `CableReport` / `ExcavationNotification` (and maybe later `PlacementContract`) and same with `Hakemus` <-- `Johtoselvitys` / `Kaivuilmoitus` (and maybe later `Sijoitussopimus`)?

Reply:
Johtoselvitys (cable report) and kaivuilmoitus (excavation notification) are not comparable terms in the application process.

Johtoselvitys is the map/document that tells the digger where the cables are underground. It's the "decision" of the process. The contractor asks for the city to produce that document. They send a johtoselvityshakemus, with which they order the johtoselvitys. The semantically correct name might be "johtoselvitystilaus", but everyone talks about applications, even the [city itself](https://www.suomi.fi/palvelut/verkkoasiointi/tilaa-johtoselvitys-tilaa-johtoselvitys-helsingin-kaupunki-kaupunkiympariston-toimiala/959df9bd-0801-4be9-9535-1b1fcdd4adc2).

Kaivuilmoitus is the application itself, which the contractor sends to the city. The city decides if the applicant has given all of the necessary information and makes a decision document. I don't think there's a special name for the decision document, so we can refer to it as a päätös, maybe kaivuilmoituspäätös or päätös kaivuilmoituksesta or kaivuilmoituksen päätös or something when we need to be more specific.

I haven't looked at placement contracts in any detail, but I think they're comparable to cable reports, i.e. sijoitussopimus is the "decision" of the process.

Reply:
Thank you for taking time to clarify this for me.

Reply:
Though it still seems that all of these different types are actually applications. Even with excavation notification (or announcement or notice or what ever), it is not just a statement that "I will dig here next week, live with it" but more like an application for a permit to dig. As you said yourself, the "**applicant**" creates the excavation notification. The result of this excavation notification would be an excavation permit. Or in Finnish it would go like "Johtoselvityshakemus" (application) --> "Johtoselvitys" (result), "Kaivuilmoitus" --> "Kaivulupa" and "Sijoitushakemus" --> "Sijoitussopimus" (maybe, not sure about this, in someplaces it is called "sijoituslupa"). But you're right about the term "Kaivuilmoitushakemus", that is wrong. The applicant does not want "kaivuilmoitus" but "kaivulupa" or "kaivulupapäätös" or something.